### ISSUE TITLE
TFLearn Memory Leak

### ISSUE THREAD

**Issue by braddengross** [tflearn/tflearn#248]
I'm running the following Flask application to train a DNN similar to AlexNet on some images. After each run I delete everything Python has available to me in memory. However, each run adds about 1GB of memory to the running Python process and produces an OOM exception sooner or later. Below is the entire application, each call to /build runs the model 30 times are random data. This is a proof of concept for a model which has far too much data to read into memory so I'm sampling the data over iterations like this. I know I can use the Image Preloader, but because I have so much data I'm actually loading pickle files in each iteration and not image files at the moment. Either way, some memory continues to be allocated over each iteration of this application which I cannot deallocate. 

`import tflearn
from flask import Flask, jsonify
from tflearn.layers.core import input_data, dropout, fully_connected
from tflearn.layers.conv import conv_2d, max_pool_2d
from tflearn.layers.normalization import local_response_normalization
from tflearn.layers.estimator import regression

import numpy as np

app = Flask(__name__)

keep_prob = .8
num_labels = 3
batch_size = 64

def reformat(dataset, labels, num_labels):
    if dataset is not None:
        dataset = dataset.reshape((-1, 227, 227, num_labels)).astype(np.float32)
    if labels is not None:
        labels = (np.arange(3) == labels[:, None]).astype(np.float32)
    return dataset, labels

class AlexNet():

    def __init__(self):

        @app.route('/build')
        def build():

            # Building 'AlexNet'
            network = input_data(shape=[None, 227, 227, 3])
            network = conv_2d(network, 96, 11, strides=4, activation='relu')
            network = max_pool_2d(network, 3, strides=2)
            network = local_response_normalization(network)
            network = conv_2d(network, 256, 5, activation='relu')
            network = max_pool_2d(network, 3, strides=2)
            network = local_response_normalization(network)
            network = conv_2d(network, 384, 3, activation='relu')
            network = conv_2d(network, 384, 3, activation='relu')
            network = conv_2d(network, 256, 3, activation='relu')
            network = max_pool_2d(network, 3, strides=2)
            network = local_response_normalization(network)
            network = fully_connected(network, 4096, activation='tanh')
            network = dropout(network, keep_prob)
            network = fully_connected(network, 4096, activation='tanh')
            network = dropout(network, keep_prob)
            network = fully_connected(network, num_labels, activation='softmax')
            network = regression(network, optimizer="adam",
                                 loss='categorical_crossentropy',
                                 learning_rate=0.001, batch_size=batch_size)

            model = tflearn.DNN(network, tensorboard_dir="./tflearn_logs/",
                                checkpoint_path=None, tensorboard_verbose=0)

            for i in range(30):
                data = np.random.randn(10, 227, 227, 3)
                labels = np.random.choice([0, 2], size=(10,), p=[1. / 3, 2. / 3])

                data, labels = reformat(data, labels, 3)

                model.fit(data, labels, n_epoch=1, shuffle=True,
                          validation_set=None,
                          show_metric=True, batch_size=batch_size, snapshot_step=None,
                          snapshot_epoch=True, run_id=None)

            del data
            del labels
            del model
            del network
            return jsonify(status=200)

if __name__ == "__main__":
    AlexNet()
    app.run(host='0.0.0.0', port=5000, threaded=True)`

**Comment by dhwajr** [comment]
try putting delete data and labels inside the for loop (30)
```
            del data
            del labels
```

**Comment by dhwajr** [comment]
Also it could be due to your fit() inside the loop. Unless you specify otherwise (using a with tf.Graph().as_default(): block), all TensorFlow operations are added to a global default graph. This means that every call to fit() adds objects to a global data structure. There are two fixes:
1. use default graph scope
```
for i in range(30):
     with tensorflow.Graph().as_default():
     .............
```
2. We can modify fit() method to not to take array inputs and rather it should accept feed to placeholders.

**Comment by aymericdamien** [comment]
As @dhwajr pointed out, every time you are going to call '/build' URL, it will append a new portion to the already existing graph. Try to encapsulate your network into a graph (such as `with tf.Graph().as_default():`).

**Comment by dhwajr** [comment]
tf.Graph().as_default helped in acheiving few more iterations of the loop, but I am experiencting memory error again. Following is the memory profiler output, something is happening inside fit() function:

```
Line #    Mem usage    Increment   Line Contents
================================================
    80 9322.402 MiB    0.000 MiB       @profile
   135 9322.402 MiB    0.000 MiB           if len(self.train_ops) == 0:
   136                                         raise Exception('tf collection "' + tf.GraphKeys.TRAIN_OPS + '" '
   137                                                         'is empty! Please make sure you are using '
   138                                                         '`regression` layer in your network.')
   139
   140 9322.402 MiB    0.000 MiB           if batch_size:
   141 9322.402 MiB    0.000 MiB               for train_op in self.train_ops:
   142 9322.402 MiB    0.000 MiB                   train_op.batch_size = batch_size
   143
   144 9322.402 MiB    0.000 MiB           valX, valY = None, None
   145 9322.402 MiB    0.000 MiB           if validation_set:
   146 9322.402 MiB    0.000 MiB               if isinstance(validation_set, float):
   147                                             valX = validation_set
   148                                             valY = validation_set
   149 9322.402 MiB    0.000 MiB               elif type(validation_set) not in [tuple, list]:
   150                                             raise ValueError("validation_set must be a tuple or list: ("
   151                                                              "valX, valY), " + str(type(validation_set))
   152                                                              + " is not compatible!")
   153                                         else:
   154 9322.402 MiB    0.000 MiB                   valX = validation_set[0]
   155 9322.402 MiB    0.000 MiB                   valY = validation_set[1]
   156
   157                                     # For simplicity we build sync dict synchronously but Trainer support
   158                                     # asynchronous feed dict allocation.
   159                                     # TODO: check memory impact for large data and multiple optimizers
   160 9322.402 MiB    0.000 MiB           feed_dict = feed_dict_builder(X_inputs, Y_targets, self.inputs,
   161 9322.402 MiB    0.000 MiB                                         self.targets)
   162 9322.402 MiB    0.000 MiB           feed_dicts = [feed_dict for i in self.train_ops]
   163 9322.402 MiB    0.000 MiB           val_feed_dicts = None
   164 9322.402 MiB    0.000 MiB           if not (is_none(valX) or is_none(valY)):
   165 9322.402 MiB    0.000 MiB               if isinstance(valX, float):
   166                                             val_feed_dicts = valX
   167                                         else:
   168 9322.402 MiB    0.000 MiB                   val_feed_dict = feed_dict_builder(valX, valY, self.inputs,
   169 9322.402 MiB    0.000 MiB                                                     self.targets)
   170 9322.402 MiB    0.000 MiB                   val_feed_dicts = [val_feed_dict for i in self.train_ops]
   171                                     # Retrieve data preprocesing and augmentation
   172 9322.402 MiB    0.000 MiB           dprep_dict, daug_dict = {}, {}
   173 9322.402 MiB    0.000 MiB           dprep_collection = tf.get_collection(tf.GraphKeys.DATA_PREP)
   174 9322.402 MiB    0.000 MiB           daug_collection = tf.get_collection(tf.GraphKeys.DATA_AUG)
   175 9322.402 MiB    0.000 MiB           for i in range(len(self.inputs)):
   176                                         # Support for custom inputs not using dprep/daug
   177 9322.402 MiB    0.000 MiB               if len(dprep_collection) > i:
   178 9322.402 MiB    0.000 MiB                   if dprep_collection[i] is not None:
   179                                                 dprep_dict[self.inputs[i]] = dprep_collection[i]
   180 9322.402 MiB    0.000 MiB               if len(daug_collection) > i:
   181 9322.402 MiB    0.000 MiB                   if daug_collection[i] is not None:
   182                                                 daug_dict[self.inputs[i]] = daug_collection[i]
   183 9322.402 MiB    0.000 MiB           self.trainer.fit(feed_dicts, val_feed_dicts=val_feed_dicts,
   184 9322.402 MiB    0.000 MiB                            n_epoch=n_epoch,
   185 9322.402 MiB    0.000 MiB                            show_metric=show_metric,
   186 9322.402 MiB    0.000 MiB                            snapshot_step=snapshot_step,
   187 9322.402 MiB    0.000 MiB                            snapshot_epoch=snapshot_epoch,
   188 9322.402 MiB    0.000 MiB                            shuffle_all=shuffle,
   189 9322.402 MiB    0.000 MiB                            dprep_dict=dprep_dict,
   190 9322.402 MiB    0.000 MiB                            daug_dict=daug_dict,
   191 9322.402 MiB    0.000 MiB                            excl_trainops=excl_trainops,
   **192 13594.859 MiB 4272.457 MiB                            run_id=run_id)**
```

**Comment by braddengross** [comment]
I was able to get around this by instantiating one graph and one session and maintaining them for the life of the application and only reloading the model on restart. However, this does not seem ideal. I believe that creating new Graphs and Sessions for each iteration and setting them as default should keep the memory stable.

**Comment by pankap** [comment]
Hi, this worked in my case to avoid the memory leak. 
```python
print('\r\nInitializing model')
network=AlexNet()
model = tflearn.DNN(network, checkpoint_path='checkpoint_' + sExperimentName,max_checkpoints=1, tensorboard_verbose=2)

# Trains 10 models in 10 folds
for fold_index in range(0,10):
    fold=Folds[fold_index]
    X = fold.TrainInputs
    Y = fold.TrainOutputs
    TX = fold.TestInputs
    TY = fold.TestOutputs

    # Prepares the names of the active experiment's fold
    sRunID = 'run_' + sExperimentName + '_%02i' % (fold_index + 1)
    sModelFileName = 'Model_' + sExperimentName + '_model_%02i.nn' % (fold_index + 1)

    # Trains the fold and saves the model
    tf.reset_default_graph()              
    with tf.Graph().as_default(), tf.Session() as sess:
        sess.run(tf.initialize_all_variables())
        tflearn.config.init_training_mode()

        print('\r\nTraining model for fold %02i' % (fold_index + 1))
        model.fit(X, Y, n_epoch=3, validation_set=(TX,TY), shuffle=False,
                  show_metric=True, batch_size=60, snapshot_step=3,
                  snapshot_epoch=False, run_id=sRunID)
        
        print('\r\nSaving model for fold %02i' % (fold_index + 1))        
        model.save(sModelFolder + sModelFileName)
```
But the model weights persisted to the next repetition and each fit() was running on an already trained model from the previous iteration. This is undesirable for my case of 10 fold cross validation, see #455.
